src/stores/feature-flag.ts

```diff
@@ -0,0 +1,75 @@
+import { defineStore } from 'pinia'
+
+import featureData from '~~/feat/feature-flags.json'
+
+import { warn } from '~/utils/console'
+
+import type { FeatureFlag } from '~/types/feature-flag'
+import type { FeatureState } from '~/constants/feature-flag'
+import { ENABLED, SWITCHABLE, ON, OFF } from '~/constants/feature-flag'
+
+export interface FeatureFlagState {
+  flags: Record<string, FeatureFlag>
+}
+
+const FEATURE_FLAG = 'feature_flag'
+
+export const useFeatureFlagStore = defineStore(FEATURE_FLAG, {
+  state: () => ({
+    flags: featureData.features as Record<string, FeatureFlag>,
+  }),
+  getters: {
+    /**
+     * Get the state of the named feature, based on config and cookie.
+     */
+    featureState:
+      (state: FeatureFlagState) =>
+      (name: string): FeatureState => {
+        if (name in state.flags) {
+          const flag = state.flags[name]
+          if (flag.status === SWITCHABLE)
+            return flag.preferredState ?? flag.defaultState
+          else if (flag.status === ENABLED) return ON
+          else return OFF
+        } else {
+          warn(`Invalid feature flag accessed: ${name}`)
+          return ON
+        }
+      },
+    /**
+     * Get the mapping of switchable features to their preferred states.
+     */
+    flagStateMap: (state: FeatureFlagState): Record<string, FeatureState> => {
+      const featureMap: Record<string, FeatureState> = {}
+      Object.entries(state.flags).forEach(([name, flag]) => {
+        if (flag.status === SWITCHABLE)
+          featureMap[name] = flag.preferredState ?? flag.defaultState
+      })
+      return featureMap
+    },
+  },
+  actions: {
+    /**
+     * Given a list of key value pairs of flags and their preferred states,
+     * populate the store state to match the cookie.
+     *
+     * @param cookies - mapping of feature flags and their preferred states
+     */
+    initFromCookies(cookies: Record<string, FeatureState>) {
+      Object.entries(this.flags).forEach(([name, flag]) => {
+        if (flag.status === SWITCHABLE) flag.preferredState = cookies[name]
+      })
+    },
+    /**
+     * Toggle the feature flag of the given name to the given preferred state.
+     *
+     * @param name - the name of the flag to toggle
+     * @param targetState - the desired state of the feature flag
+     */
+    toggleFeature(name: string, targetState: FeatureState) {
```

Comment:
I see :thinking: Why is that? The store already knows how to initialize itself from cookies. Is there a usecase where we'd want to call `toggleFeature` and it _wouldn't_ update the cookie?

Reply:
The store doesn't read cookie directly it only reads the JSON (which is read from cookies in middleware). The store also doesn't write cookies, it just returns an object (which is written to the cookie in a Vue component).

Reply:
Gotcha!

I wonder of toggle feature is a confusing function expose then? Couldn't the "save" function on the preferences page just pass the updated cookie JSON to the `init` function? What's the use-case for this function outside of the preferences page? In the current implementation you could have some confusing behavior because the preferences page actually does toggle the feature in the store without updating the cookie, meaning if you navigate away from the preferences page without reloading then the rest of your session will be feature flags that are out of sync with your cookies.

To me it feels like splitting the store from the cookie introduces two sources of truth that can very easily drift (and indeed do in the currently implementation).

Reply:
Technically the source of truth is always the store, 100% of the time. It's the only place from which the feature state is determined. The cookie is only meant for persistence of a very small subset of the info held in the store, i.e. mapping of switchable flags to their preferred states.

Reply:
Yup! Sounds good to me. As long as can make sure the cookies and the store state don't drift (so that the unexpected "reload" with suddenly different features than you expected based on the previous page's state doesn't happen) then it's good to me!